cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/SchemaParser.ts

```diff
@@ -0,0 +1,431 @@
+  const t = getNonNullableType(type);
+  const id = uuidV4();
+  if (t.items && !isComplexType(t.items)) {
+    return {
+      [id]: {
+        internalType: InternalTypesEnum.ARRAY_SIMPLE_TYPE,
+        id,
+        nullable: isNullable(t.items),
+        type: getNonNullableType(t.items),
+      },
+    };
+  }
+  return {
+    [id]: {
+      internalType: InternalTypesEnum.ARRAY_COMPLEX_TYPE_ROOT,
+      id,
+      nullable,
+      type: getComplexTypeName(t.items),
+      children: parseComplexType(t.items),
+    },
+  };
+}
+
+/**
+ * @returns
+ * {
+ *  order: [child-id1, child-id2,...]
+ *  [child-id1]: {
+ *    id: child-id1,
+ *    internalType: 'enum-symbol',
+ *    typeProperties: {
+ *      symbol: 'symbol1'
+ *    }
+ *  },
+ *  [child-id2]: {
+ *    id: child-id2,
+ *    internalType: 'enum-symbol',
+ *    typeProperties: {
+ *      symbol: 'symbol2'
+ *    }
+ *  }
+ * }
+ * @param type avro enum type
+ * {
+ *  type: 'enum',
+ *  symbols: ['symbol1', 'symbol2', ....]
+ * }
+ */
+function parseEnumType(type): IOrderedChildren {
+  const nullable = isNullable(type);
+  const t = getNonNullableType(type);
+  const result = {
+    order: [],
+  };
+  for (const symbol of t.symbols) {
+    const id = uuidV4();
+    result.order.push(id);
+    result[id] = {
+      id,
+      internalType: InternalTypesEnum.ENUM_SYMBOL,
+      nullable,
+      typeProperties: {
+        symbol,
+      },
+    };
+  }
+  return result;
+}
+
+function getMapSubType(type, internalTypeName): INode {
+  const id = uuidV4();
+  if (!isComplexType(type)) {
+    return {
+      id,
+      internalType: internalTypeName.simpleType,
+      nullable: isNullable(type),
+      type: getNonNullableType(type),
+    };
+  } else {
+    const complexType = getComplexTypeName(type);
+    const nullable = isNullable(type);
+    return {
+      children: parseComplexType(type),
+      id,
+      internalType: internalTypeName.complexType,
+      type: complexType,
+      nullable,
+    };
+  }
+}
+/**
+ * @returns
+ * {
+ *   [child-id1]: {
+ *    "id": child-id1,
+ *    "internalType": "map-keys-simple-type",
+ *    "type": "string"
+ *  },
+ *   [child-id2] {
+ *    "id": child-id2,
+ *    "internalType": "map-values-simple-type",
+ *    "nullable": false,
+ *    "type": "string"
+ *   }
+ * }
+ * @param type avro map type
+ * {
+ *  type: {
+ *    keys: 'string',
+ *    values: 'string',
+ *  }
+ * }
+ */
+function parseMapType(type): IOrderedChildren {
+  const t = getNonNullableType(type);
+  const keysType = t.keys;
+  const valuesType = t.values;
+  const result: Record<string, INode> = {};
+  const mapKeysSubType = getMapSubType(keysType, {
+    simpleType: InternalTypesEnum.MAP_KEYS_SIMPLE_TYPE,
+    complexType: InternalTypesEnum.MAP_KEYS_COMPLEX_TYPE_ROOT,
+  });
+  const mapValuesSubType = getMapSubType(valuesType, {
+    simpleType: InternalTypesEnum.MAP_VALUES_SIMPLE_TYPE,
+    complexType: InternalTypesEnum.MAP_VALUES_COMPLEX_TYPE_ROOT,
```

Comment:
do these need to be exposed here? can these value just be inside the `getMapSubType`?

Reply:
I am using it for both keys and values type. Depending on the usage the simpleType and complexType differs. Do you recommend anything specific we can do here?